pkg/graveler/graveler.go

```diff
@@ -1851,8 +1846,51 @@ func (g *KVGraveler) stagingEmpty(ctx context.Context, repositoryID RepositoryID
 	return true, nil
 }
 
+// dropStaging deletes all staging area entries of a given branch from store
+// We do not wait for result as this is an asynchronous operation
+func (g *KVGraveler) dropStaging(ctx context.Context, branch *Branch) {
+	tokens := branch.SealedTokens
+	tokens = append(tokens, branch.StagingToken)
+	for _, st := range tokens {
+		token := st // pinning
+		go func() {
+			err := g.StagingManager.Drop(ctx, token)
+			if err != nil {
+				logging.Default().WithError(err).WithField("staging_token", token).Error("failed to drop staging token")
+			}
+		}()
+	}
+}
+
 func (g *KVGraveler) Reset(ctx context.Context, repositoryID RepositoryID, branchID BranchID) error {
-	return g.db.Reset(ctx, repositoryID, branchID)
+	isProtected, err := g.protectedBranchesManager.IsBlocked(ctx, repositoryID, branchID, BranchProtectionBlockedAction_STAGING_WRITE)
+	if err != nil {
+		return err
+	}
+	if isProtected {
+		return ErrWriteToProtectedBranch
+	}
+	currBranch, err := g.RefManager.GetBranch(ctx, repositoryID, branchID)
+	if err != nil {
+		return err
+	}
```

Comment:
Use the branch given to you by the `BranchUpdate` func. The value here may be outdated.
```suggestion
```